### DISCUSSION TITLE
[Question]mfspreflabels = LABELEXPR

### DISCUSSION THREAD

**Discussion by Sveeeeeen** [moosefs/moosefs#393] [Q&A]
Hello,

According to the storage class manual we can force mfsmount to prefer specific labels over others.
It's not explained though how LABELEXPR should looks like or am I wrong?

> Preferred labels in MooseFS Client are a list (up to 9) of labels expressions, e.g. E1, E2, E3.

For example, I have 3 Labels, A, B and C.

`mfspreflabels=A,B,C`
Is not accepted by mfsmount. How it should looks like?

What I am trying to achieve is to prefer A over B and B over C. A > B > C. Like a waterfall.
Is this case possible? 

Thanks in advance!
Sven

**Comment by pkonopelko** [comment]
Hi @Sveeeeeen,

> What I am trying to achieve is to prefer A over B and B over C. A > B > C. Like a waterfall.

This is exactly how it should work. The correct way to use this parameter is presented below. If more than one value is given, single quote marks are needed:

```
-o mfspreflabels='A,B,C'
```

You may check if it mounted correctly and `preflabels` are respected by running: `cat /mnt/mfs/.params` (assuming `/mnt/mfs` is a rootdir of your MooseFS mounpoint).

`.params` file is not a real file, therefore it is not visible with `ls`, even with `-a` option. Just try to read it and you should see a set of current mounting options.

Let me know if that worked for you.
Piotr

**Comment by Sveeeeeen** [comment]
Thanks for the fast response!

I am using the  mfscfgfile=/etc/mfs/mfsmount.cfg paramter.

I tried 
`mfspreflabels='A,B,C'`
Inside mfsmount.cfg right now and I get a parsing error.

> parse error, next char: '''
error parsing preferred labels expression
see: mfsmount -h for help

**Comment by pkonopelko** [comment]
In `mfsmount.cfg` please use the following line:

```
mfspreflabels=A;B;C
```

Piotr

**Comment by Sveeeeeen** [comment]
Thank you very much!


> cat /mnt/mfsmount/.params | grep pref
> mfspreflabels: A;B;C

Looks much better now!

**Comment by eleaner** [comment]
@pkonopelko 
I don't seem to be able to set the preferred labels (except a single label)
neither through fstab nor mfsmount

this would be my preferred access sequence
1. low latency SSD - taggetd T,S
2. low(er) latency HDD - tagged T,R
3. other SSD - tagged S
4. other HDD - tagged R

mfspreflabels=T+S,T+R,S,R
except comma in fstab indicates the start of the next option, neither single or double inverted commas seem to work

please help
mfsmount -V
MFS version: 4.57.6-1-pro ; build: 1944
using FUSE kernel interface version 7.31

**Reply by eleaner** [reply]
unless the correct sequence is T+S;T+R;S;R

so -o mfspreflabels="T+S;T+R;S;R" in mfsmount
or ....,mfspreflabels=T+S;T+R;S;R,.... in fstab
?

**Reply by chogata** [reply]
`-o mfspreflabels="T+S;T+R;S;R"` or  `-o mfspreflabels="T+S\,T+R\,S\,R"` in command line mount (man mentions only `;` in the label grammar because of that, but escaping commas works too)
 `mfspreflabels=T+S;T+R;S;R` in fstab
